File: textiles_and_garments/textiles_and_garments/report/po___greige_awaiting_goods_in_nap/po___greige_awaiting_goods_in_nap.js
// // Copyright (c) 2026, Aravind and contributors
// // For license information, please see license.txt

// frappe.query_reports["PO - Yarn Awaiting goods in NAP"] = {
// 	"filters": [
// 		{
// 			"fieldname": "company",
// 			"label": __("Company"),
// 			"fieldtype": "Link",
// 			"options": "Company",
// 			"default": frappe.defaults.get_user_default("Company"),
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "supplier",
// 			"label": __("Supplier"),
// 			"fieldtype": "Link",
// 			"options": "Supplier",
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "from_date",
// 			"label": __("From Date"),
// 			"fieldtype": "Date",
// 			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -3),
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "to_date",
// 			"label": __("To Date"),
// 			"fieldtype": "Date",
// 			"default": frappe.datetime.get_today(),
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "item_code",
// 			"label": __("Item Code"),
// 			"fieldtype": "Link",
// 			"options": "Item",
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "material_request",
// 			"label": __("Material Request"),
// 			"fieldtype": "Link",
// 			"options": "Material Request",
// 			"reqd": 0
// 		}
// 	]
// };

// Copyright (c) 2026, Aravind and contributors
// For license information, please see license.txt

// // Copyright (c) 2026, Aravind and contributors
// // For license information, please see license.txt

// frappe.query_reports["PO - Yarn Awaiting goods in NAP"] = {
// 	"filters": [
// 		{
// 			"fieldname": "company",
// 			"label": __("Company"),
// 			"fieldtype": "Link",
// 			"options": "Company",
// 			"default": frappe.defaults.get_user_default("Company"),
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "supplier",
// 			"label": __("Supplier"),
// 			"fieldtype": "Link",
// 			"options": "Supplier",
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "from_date",
// 			"label": __("From Date"),
// 			"fieldtype": "Date",
// 			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -3),
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "to_date",
// 			"label": __("To Date"),
// 			"fieldtype": "Date",
// 			"default": frappe.datetime.get_today(),
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "item_code",
// 			"label": __("Item Code"),
// 			"fieldtype": "Link",
// 			"options": "Item",
// 			"reqd": 0
// 		},
// 		{
// 			"fieldname": "material_request",
// 			"label": __("Material Request"),
// 			"fieldtype": "Link",
// 			"options": "Material Request",
// 			"reqd": 0
// 		}
// 	]
// };

// Copyright (c) 2026, Aravind and contributors
// For license information, please see license.txt

frappe.query_reports["PO - Greige Awaiting goods in NAP"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 0
		},
		{
			"fieldname": "supplier",
			"label": __("Supplier"),
			"fieldtype": "Link",
			"options": "Supplier",
			"reqd": 0
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -3),
			"reqd": 0
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 0
		},
		{
			"fieldname": "item_code",
			"label": __("Item Code"),
			"fieldtype": "Link",
			"options": "Item",
			"reqd": 0
		},
		{
			"fieldname": "material_request",
			"label": __("Material Request"),
			"fieldtype": "Link",
			"options": "Material Request",
			"reqd": 0
		}
	],
	
	"onload": function(report) {
		// Add Summary button
		report.page.add_inner_button(__("Summary"), function() {
			let data = report.data;
			
			if (!data || data.length === 0) {
				frappe.msgprint(__("No data available to summarize"));
				return;
			}
			
			// Filter out total rows (rows without purchase_order or with specific indicators)
			let filtered_data = data.filter(row => {
				// Exclude rows that are totals/subtotals
				// Total rows typically don't have a purchase_order value or have specific markers
				return row.purchase_order && 
				       !row.is_total_row && 
				       row.purchase_order !== "Total" &&
				       row.purchase_order !== "Grand Total";
			});
			
			if (filtered_data.length === 0) {
				frappe.msgprint(__("No valid data rows found"));
				return;
			}
			
			// Calculate summary statistics
			let total_records = filtered_data.length;
			let unique_suppliers = [...new Set(filtered_data.map(row => row.supplier).filter(Boolean))].length;
			let unique_pos = [...new Set(filtered_data.map(row => row.purchase_order).filter(Boolean))].length;
			
			let total_ordered = filtered_data.reduce((sum, row) => sum + (parseFloat(row.qty) || 0), 0);
			let total_received = filtered_data.reduce((sum, row) => sum + (parseFloat(row.received_qty) || 0), 0);
			let total_pending = filtered_data.reduce((sum, row) => sum + (parseFloat(row.pending_qty) || 0), 0);
			
			let avg_per_received = filtered_data.reduce((sum, row) => sum + (parseFloat(row.per_received) || 0), 0) / total_records;
			let avg_aging = filtered_data.reduce((sum, row) => sum + (parseFloat(row.aging_days) || 0), 0) / total_records;
			
			// Show summary dialog
			frappe.msgprint({
				title: __("Greige Awaiting GRN - Summary"),
				message: `
					<div style="font-size: 14px; line-height: 2;">
						<h4 style="margin-bottom: 15px; color: #2563eb;">📊 Report Overview</h4>
						<table style="width: 100%; border-collapse: collapse;">
							<tr style="border-bottom: 1px solid #e5e7eb;">
								<td style="padding: 8px 0;"><strong>Total Line Items:</strong></td>
								<td style="text-align: right; padding: 8px 0;">${total_records}</td>
							</tr>
							<tr style="border-bottom: 1px solid #e5e7eb;">
								<td style="padding: 8px 0;"><strong>Unique Purchase Orders:</strong></td>
								<td style="text-align: right; padding: 8px 0;">${unique_pos}</td>
							</tr>
							<tr style="border-bottom: 1px solid #e5e7eb;">
								<td style="padding: 8px 0;"><strong>Unique Suppliers:</strong></td>
								<td style="text-align: right; padding: 8px 0;">${unique_suppliers}</td>
							</tr>
						</table>
						
						<h4 style="margin: 20px 0 15px 0; color: #16a34a;">📦 Quantity Summary</h4>
						<table style="width: 100%; border-collapse: collapse;">
							<tr style="border-bottom: 1px solid #e5e7eb;">
								<td style="padding: 8px 0;"><strong>Total Ordered Qty:</strong></td>
								<td style="text-align: right; padding: 8px 0;">${total_ordered.toFixed(2)} Kgs</td>
							</tr>
							<tr style="border-bottom: 1px solid #e5e7eb;">
								<td style="padding: 8px 0;"><strong>Total Received Qty:</strong></td>
								<td style="text-align: right; padding: 8px 0; color: #0284c7;">${total_received.toFixed(2)} Kgs</td>
							</tr>
							<tr style="border-bottom: 1px solid #e5e7eb;">
								<td style="padding: 8px 0;"><strong>Total Pending Qty:</strong></td>
								<td style="text-align: right; padding: 8px 0; color: #dc2626; font-weight: 700;">${total_pending.toFixed(2)} Kgs</td>
							</tr>
						</table>
						
						<h4 style="margin: 20px 0 15px 0; color: #ea580c;">📈 Performance Metrics</h4>
						<table style="width: 100%; border-collapse: collapse;">
							<tr style="border-bottom: 1px solid #e5e7eb;">
								<td style="padding: 8px 0;"><strong>Average % Received:</strong></td>
								<td style="text-align: right; padding: 8px 0;">${avg_per_received.toFixed(2)}%</td>
							</tr>
							<tr style="border-bottom: 1px solid #e5e7eb;">
								<td style="padding: 8px 0;"><strong>Average Aging:</strong></td>
								<td style="text-align: right; padding: 8px 0;">${avg_aging.toFixed(1)} Days</td>
							</tr>
						</table>
					</div>
				`,
				indicator: "blue",
				wide: true
			});
		});
	}
};

File: textiles_and_garments/textiles_and_garments/report/po___greige_awaiting_goods_in_nap/po___greige_awaiting_goods_in_nap.py
# Copyright (c) 2026, Aravind and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data

def get_columns():
	"""Define report columns"""
	return [
		{
			"fieldname": "purchase_order",
			"label": _("Purchase Order"),
			"fieldtype": "Link",
			"options": "Purchase Order",
			"width": 150
		},
		{
			"fieldname": "transaction_date",
			"label": _("PO Date"),
			"fieldtype": "Date",
			"width": 100
		},
		{
			"fieldname": "supplier",
			"label": _("Supplier"),
			"fieldtype": "Link",
			"options": "Supplier",
			"width": 150
		},
		{
			"fieldname": "supplier_name",
			"label": _("Supplier Name"),
			"fieldtype": "Data",
			"width": 180
		},
		{
			"fieldname": "status",
			"label": _("Status"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "item_code",
			"label": _("Item Code"),
			"fieldtype": "Link",
			"options": "Item",
			"width": 150
		},
		{
			"fieldname": "item_name",
			"label": _("Item Name"),
			"fieldtype": "Data",
			"width": 200
		},
		{
			"fieldname": "qty",
			"label": _("Ordered Qty"),
			"fieldtype": "Float",
			"width": 100
		},
		{
			"fieldname": "stock_uom",
			"label": _("UOM"),
			"fieldtype": "Link",
			"options": "UOM",
			"width": 80
		},
		{
			"fieldname": "received_qty",
			"label": _("Received Qty"),
			"fieldtype": "Float",
			"width": 100
		},
		{
			"fieldname": "pending_qty",
			"label": _("Pending Qty"),
			"fieldtype": "Float",
			"width": 100
		},
		{
			"fieldname": "per_received",
			"label": _("% Received"),
			"fieldtype": "Percent",
			"width": 100
		},
		{
			"fieldname": "schedule_date",
			"label": _("Schedule Date"),
			"fieldtype": "Date",
			"width": 110
		},
		{
			"fieldname": "shipping_address",
			"label": _("Shipping Address"),
			"fieldtype": "Data",
			"width": 200
		},
		{
			"fieldname": "material_request",
			"label": _("Material Request"),
			"fieldtype": "Link",
			"options": "Material Request",
			"width": 150
		},
		{
			"fieldname": "aging_days",
			"label": _("Aging (Days)"),
			"fieldtype": "Int",
			"width": 100
		},
		{
			"fieldname": "company",
			"label": _("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"width": 200
		}
	]

def get_data(filters):
	"""
	Fetch Purchase Order items with conditions:
	- Document Status = Submitted
	- Item Code LIKE '%yr%' (yarn items)
	- % Received NOT EQUALS 100
	- % Received LESS THAN 95
	- Shipping Address LIKE '%NAP%'
	"""
	conditions = get_conditions(filters)
	
	data = frappe.db.sql(f"""
		SELECT 
			po.name as purchase_order,
			po.transaction_date,
			po.supplier,
			po.supplier_name,
			po.status,
			po.company,
			po.shipping_address as shipping_address,
			poi.item_code,
			poi.item_name,
			poi.qty,
			poi.stock_uom,
			poi.received_qty,
			(poi.qty - poi.received_qty) as pending_qty,
			ROUND((poi.received_qty / poi.qty * 100), 2) as per_received,
			poi.schedule_date,
			poi.material_request,
			DATEDIFF(CURDATE(), po.transaction_date) as aging_days
		FROM 
			`tabPurchase Order` po
		INNER JOIN 
			`tabPurchase Order Item` poi ON poi.parent = po.name
		WHERE 
			po.docstatus = 1
			AND poi.item_code LIKE '%%GKF%%'
			AND ROUND((poi.received_qty / poi.qty * 100), 2) != 100.00
			AND ROUND((poi.received_qty / poi.qty * 100), 2) < 95.00
			AND po.shipping_address LIKE '%%NAP%%'
			{conditions}
		ORDER BY 
			po.transaction_date DESC, po.name, poi.idx
	""", filters, as_dict=1)
	
	return data

def get_conditions(filters):
	"""Build additional filter conditions"""
	conditions = ""
	
	if filters.get("company"):
		conditions += " AND po.company = %(company)s"
	
	if filters.get("supplier"):
		conditions += " AND po.supplier = %(supplier)s"
	
	if filters.get("from_date"):
		conditions += " AND po.transaction_date >= %(from_date)s"
	
	if filters.get("to_date"):
		conditions += " AND po.transaction_date <= %(to_date)s"
	
	if filters.get("item_code"):
		conditions += " AND poi.item_code = %(item_code)s"
	
	if filters.get("material_request"):
		conditions += " AND poi.material_request = %(material_request)s"
	
	return conditions